h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -497,32 +593,45 @@ public MP nextModelParameters(Model previousModel) {
-            MP commonModelParams = (MP) _params.clone();
-            // Fill model parameters
-            MP params = getModelParams(commonModelParams, hypers);
-
-            // add max_runtime_secs in search criteria into params if applicable
-            if (_search_criteria != null && _search_criteria.strategy() == HyperSpaceSearchCriteria.Strategy.RandomDiscrete) {
-              // ToDo: model seed setting will be different for parallel model building.
-              // ToDo: This implementation only works for sequential model building.
-              if (_set_model_seed_from_search_seed) {
-                // set model seed = search_criteria.seed+(0, 1, 2,..., model number)
-                params._seed = _search_criteria.seed() + (model_number++);
-              }
+            Object[] permutation = permutation(_currentHyperparamIndices, new Object[_hyperParamNames.length]);
+             
+            if (_filterFunctions == null || !permutationIsSkipped(permutation, _hyperParamNames, _filterFunctions)) {
+
+              _numberOfUsedPermutations++;
+              // Get clone of parameters
+              MP commonModelParams = (MP) _params.clone();
+              // Fill model parameters
+              MP params = getModelParams(commonModelParams, permutation);
+
+              // add max_runtime_secs in search criteria into params if applicable
+              if (_search_criteria != null && _search_criteria.strategy() == HyperSpaceSearchCriteria.Strategy.RandomDiscrete) {
+                // ToDo: model seed setting will be different for parallel model building.
+                // ToDo: This implementation only works for sequential model building.
+                if (_set_model_seed_from_search_seed) {
+                  // set model seed = search_criteria.seed+(0, 1, 2,..., model number)
+                  params._seed = _search_criteria.seed() + (model_number++);
+                }
 
-              // set max_runtime_secs
-              double timeleft = this.time_remaining_secs();
-              if (timeleft > 0)  {
-                if (params._max_runtime_secs > 0) {
-                  params._max_runtime_secs = min(params._max_runtime_secs, timeleft);
-                } else {
-                  params._max_runtime_secs = timeleft;
+                // set max_runtime_secs
+                double timeleft = this.time_remaining_secs();
+                if (timeleft > 0) {
+                  if (params._max_runtime_secs > 0) {
+                    params._max_runtime_secs = min(params._max_runtime_secs, timeleft);
+                  } else {
+                    params._max_runtime_secs = timeleft;
+                  }
                 }
               }
+              return params;
+            } else {
+              
+              int nextPermutationNum = _currentPermutationNum + 1;
+              if(hasNext(previousModel) && (max_models() == 0 || max_models() > 0 && nextPermutationNum < max_models())) {
```

Comment:
why the `(max_models() == 0 || max_models() > 0 && nextPermutationNum < max_models())` ?
This condition should stay external to the original call to `nextModelParameters`. The later method is only concerned with providing the next set of params, not about stopping conditions.

Reply:
which makes me wonder if the distinction between `_numberOfUsedPermutations` and `_currentPermutationNum` is actually useful.

Reply:
> The later method is only concerned with providing the next set of params, not about stopping conditions.

If we step outside of the `nextModelParameters` , then we loose time for instantiating ModelParameters that are not actually needed. Once we don't want to step outside - we need to check for early stopping here as well.

`_currentPermutationNum` is not actually being used anywhere outside of the iterator. I was considering to substitute it with `_numberOfUsedPermutations`. A final decision was to not to touch existing variable as purpose and design was unclear to me. But let's change this as well.